testonly/integration/clientserver.go

```diff
@@ -91,87 +99,145 @@ func getTestDB(testID string) (*sql.DB, error) {
 	}
 	for _, sql := range resetSQL {
 		if _, err := dbRoot.Exec(sql); err != nil {
-			return nil, err
+			return nil, "", err
 		}
 	}
 
 	// Create new database.
 	dbTest, err := sql.Open("mysql", testDBURI)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 	createSQL, err := ioutil.ReadFile(createSQLFile)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 	sqlSlice := strings.Split(string(createSQL), ";\n")
 	// Omit the last element of the slice, since it will be "".
 	for _, sql := range sqlSlice[:len(sqlSlice)-1] {
 		if _, err := dbTest.Exec(sql); err != nil {
-			return nil, err
+			return nil, "", err
 		}
 	}
 
-	return dbTest, nil
+	return dbTest, testDBURI, nil
 }
 
-// NewLogEnv creates a fresh DB, log server, and client.
+// NewLogEnv creates a fresh DB, log server, and client. The numSequencers parameter
+// indicates how many sequencers to run in parallel; if numSequencers is zero a
+// manually-controlled test sequencer is used.
 // testID should be unique to each unittest package so as to allow parallel tests.
-func NewLogEnv(testID string) (*LogEnv, error) {
-	db, err := getTestDB(testID)
+func NewLogEnv(ctx context.Context, numSequencers int, testID string) (*LogEnv, error) {
+	var env LogEnv
+	var err error
+	env.DB, env.dbURI, err = getTestDB(testID)
 	if err != nil {
 		return nil, err
 	}
 
 	timesource := &util.SystemTimeSource{}
-	registry, err := builtin.NewDefaultExtensionRegistry()
+	registry, err := builtin.NewExtensionRegistry(env.dbURI)
 	if err != nil {
 		return nil, err
 	}
 
-	// Start Log Server.
-	grpcServer := grpc.NewServer()
-	logServer := server.NewTrillianLogRPCServer(registry, timesource)
-	trillian.RegisterTrillianLogServer(grpcServer, logServer)
+	// Create Log Server.
+	env.grpcServer = grpc.NewServer()
+	env.logServer = server.NewTrillianLogRPCServer(registry, timesource)
+	trillian.RegisterTrillianLogServer(env.grpcServer, env.logServer)
 
-	// Start Sequencer.
+	// Create Sequencer.
 	keyManager, err := crypto.LoadPasswordProtectedPrivateKey(privateKeyFile, privateKeyPassword)
 	if err != nil {
 		return nil, err
 	}
-	sequencerManager := server.NewSequencerManager(keyManager, registry,
+	env.LogOperation = server.NewSequencerManager(keyManager, registry,
 		sequencerWindow)
-	sequencerTask := server.NewLogOperationManagerForTest(ctx, registry,
-		batchSize, sleep, timeSource, sequencerManager)
+	if numSequencers == 0 {
```

Comment:
That's not how it works -- the existing code is somewhat confusing because it gives the name `sequencerTask` to a `LogOperationManager` object (of which there is only one).  The `LogOperationManager` will then spawn goroutines for the given count of sequencers.